File: Auditory/4004.js
var scanf = require('scanf');


var n = scanf('%d');

var matrix = [];
for (let i = 0; i < n; i++) {
    matrix.push(scanf(('%d '.repeat(n)).trim()));
}

function hasCycle(n, matrix) {
    const visited = Array(n).fill(0);

    function dfs(v) {
        visited[v] = 1;
        for (let u = 0; u < n; u++) {
            if (matrix[v][u]) {
                if (visited[u] === 1) return true;
                if (visited[u] === 0 && dfs(u)) return true;
            }
        }
        visited[v] = 2;
        return false;
    }

    for (let i = 0; i < n; i++) {
        if (visited[i] === 0 && dfs(i)) {
            console.log(1);
            return;
        }
    }
    console.log(0);
}

hasCycle(n, matrix);

File: Auditory/4004.py
def has_cycle(n, matrix):
    visited = [0] * n

    def dfs(v):
        visited[v] = 1
        for u in range(n):
            if matrix[v][u]:
                if visited[u] == 1:
                    return True
                if visited[u] == 0 and dfs(u):
                    return True
        visited[v] = 2
        return False

    for i in range(n):
        if visited[i] == 0:
            if dfs(i):
                return 1
    return 0

n = int(input())
matrix = [list(map(int, input().split())) for _ in range(n)]
print(has_cycle(n, matrix))